p2p/peers.go

```diff
@@ -111,18 +111,100 @@ func (n *QriNode) PeerInfo(pid peer.ID) pstore.PeerInfo {
+// Since we want to limit the number of calls to the registry,
+// only check the reputation if the given reputation has not changed
+// from our baseline reputation. If it has changed, we know we have
+// looked up the reputation before, as the registry must respond with
+// a non zero number.
+func (n *QriNode) checkReputation(profile *profile.Profile) error {
+	pids := profile.PeerIDs
+	reps := []int{}
+	// assume we do not have to disconnect
+	disconnect := false
+	// assume we have already checked this
+	checked := true
+	// check to see if this reputation has been checked before
+	// get the reputation of each peerID from the peerstore
+	for _, pid := range pids {
+		_rep, err := n.Host.Peerstore().Get(pid, qriReputationKey)
+		if err != nil {
+			if err != pstore.ErrNotFound {
+				return fmt.Errorf("error getting peer's reputation from the peerstore: %s", err.Error())
+			}
+			// if the reputation is not in the peerstore, tag it with the qriBaseReputation
+			if err := n.TagReputation(pid, qriBaselineReputation); err != nil {
+				return fmt.Errorf("error tagging connection: %s", err)
+			}
+		}
+		rep := qriBaselineReputation
+		if _rep != nil {
+			rep = _rep.(int)
+		}
+		// if the reputation has not been changed, then it has not been checked
+		if rep == qriBaselineReputation {
+			checked = false
+		}
+		if rep < 0 {
+			disconnect = true
+		}
+		reps = append(reps, rep)
+	}
+
+	// if it has been checked already, return
+	if checked {
+		// if the reputation is negative, disconnect
+		if disconnect {
+			n.DisconnectFromProfile(profile)
+			return nil
+		}
+		return nil
+	}
+
+	// create a regclient and get the reputation from the registry
+	loc := n.RegistryLocation()
+	if loc == "" {
+		return fmt.Errorf("check reputation: no registry provided")
+	}
+	cfg := &regclient.Config{Location: loc}
+	client := regclient.NewClient(cfg)
+	reputation, err := client.GetReputation(profile.ID.String())
+	if err != nil {
+		return err
+	}
+	// get integer value of reputation
+	rep := reputation.Reputation()
+	// for each peerID, change the tag value and maybe close connection:
+	for i, pid := range pids {
+		newRep := reps[i] + rep
+		// put the new reputation in the peerstore & conn manager
+		if err := n.TagReputation(pid, newRep); err != nil {
+			return fmt.Errorf("error tagging connection: %s", err)
+		}
+		// close connection if value is negative
+		if newRep < 0 {
+			disconnect = true
+		}
+	}
+	if disconnect {
+		n.DisconnectFromProfile(profile)
+	}
+	return nil
+}
+
 // AddQriPeer negotiates a connection with a peer to get their profile details
-// and peer list.
+// and peer list. Checks reputation, updates conn manager with reputation, closes
+// connection if resulting reputation is negative.
 func (n *QriNode) AddQriPeer(pinfo pstore.PeerInfo) error {
-	// add this peer to our store so libp2p has the provided addresses of
-	// the peer in the next call
-	n.Host.Peerstore().AddAddrs(pinfo.ID, pinfo.Addrs, pstore.TempAddrTTL)
```

Comment:
Okay. So. Having a hard time with this one line. 

the `p2ptest.ConnectQriPeers` is used throughout our codebase as the way to connect nodes. It basically wraps `AddPeers` which wraps `AddQriPeers`. However, AddQriPeers doesn't do anything with connecting two peers, it just on-boards a peer if it is a Qri peer, after we have established a connection. 

This line here, where we add the peerinfo to the AddressBook is why `p2ptest.ConnectQriPeers` works. To connect two qri peers, we should really be calling `n.Host.Connect`, which takes care of adding the Addrs to the address book. However, I'm not totally sure how much we want to add to our `TestablePeerNode` interface and what the logic behind keeping it small might be. Another option would be to call `n.Host.Connect(ctx, peerinfo)` in`n.AddPeer` before calling `AddQriPeer`. 

Basically, removing this line breaks all the tests that use `p2p.ConnectQriPeers`, and I'm struggling with figuring out the best way to resolve.

Reply:
First off, apologies because the whole situation with ConnectNodes and ConnectQriPeers is messy and needs a major overall. I'll describe what I think the current reality is, but I may have certain details incorrect (I'm still constantly confused by it, and had to just read through enough of the code to remember what's going on).

> the p2ptest.ConnectQriPeers is used throughout our codebase as the way to connect nodes

This is not accurate. `p2ptest.ConnectNodes` actually creates the live communication connection channel between nodes. `p2ptest.ConnectQriPeers` is doing something else.

The problem is that we have at least 3 meanings of the term "connect" right now:

* libp2p connect - This is `Host.Connect` and `Network.DialPeer`, the thing you're talking about; a live communication channel between nodes.
* peer list & profile - This is what `p2p.ConnectQriPeers` is doing right now. It's trading profile data between nodes. Most of our p2p tests are depending on this happening right now, but don't actually care about network connection.
* mapstore's mock ipfs links - This is call to `m0.AddConnection` seen in some tests, a way to immitate ipfs's global data so that filenames work across nodes

My vote is to modify `p2ptest.ConnectQriPeers` as little as possible right now, and definitely don't change its semantics. If the reputation implementation needs nodes to be connected over a communication channel, it should be using `p2ptest.ConnectNodes` instead. Then, in the near future, I'll try making a change that cleans all of this up by having a single method "connect" nodes, with an options struct defining exactly what kind of connection we need (libp2p, peer&profile, mock ifps, others...).